firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/PatchMutation.java

```diff
@@ -142,6 +143,32 @@ public void applyToLocalView(MutableDocument document, Timestamp localWriteTime)
         .setHasLocalMutations();
   }
 
+  @Override
+  public Mutation squash(
+      Mutation baseMutation, MutableDocument document, Timestamp localWriteTime) {
+    if (!getPrecondition().isValidFor(document)) {
+      return baseMutation;
+    }
+
+    Map<FieldPath, Value> transformResults = localTransformResults(localWriteTime, document);
+
+    ObjectValue mergedPatch =
+        baseMutation.getValue() != null ? baseMutation.getValue() : new ObjectValue();
+    mergedPatch.setAll(getPatch());
+    mergedPatch.setAll(transformResults);
+
+    if (baseMutation.getMask() == null) {
+      // TODO: Use strictest preconditions or assert we only use exists
+      return new SetMutation(getKey(), mergedPatch, getPrecondition());
+    } else {
+      HashSet<FieldPath> mergedMaskSet = new HashSet<>(baseMutation.getMask().getMask());
+      mergedMaskSet.addAll(mask.getMask());
```

Comment:
`mergedMaskSet` should also include fields from transforms.

Reply:
That is correct.